core/src/main/java/com/graphhopper/routing/AStarBidirection.java

```diff
@@ -61,272 +53,95 @@
-        pqOpenSetTo.add(currTo);
-
-        if (currFrom != null) {
-            currFrom.weight += weightApprox.approximate(currFrom.adjNode, false);
-            currTo.weight += weightApprox.approximate(currTo.adjNode, true);
-        }
-
-        if (!traversalMode.isEdgeBased()) {
-            bestWeightMapTo.put(to, currTo);
-            if (currFrom != null) {
-                bestWeightMapOther = bestWeightMapFrom;
-                updateBestPath(GHUtility.getEdge(graph, currFrom.adjNode, to), currFrom, to);
-            }
-        } else if (currFrom != null && currFrom.adjNode == to) {
-            // special case of identical start and end
-            bestPath.sptEntry = currFrom;
-            bestPath.edgeTo = currTo;
-            finishedFrom = true;
-            finishedTo = true;
-        }
+        super.init(from, fromWeight, to, toWeight);
     }
 
     @Override
-    protected Path createAndInitPath() {
-        bestPath = new PathBidirRef(graph, weighting);
-        return bestPath;
+    protected boolean finished() {
+        // using 'weight' is important and correct here e.g. approximation can get negative and smaller than 'weightOfVisitedPath'
+        return super.finished();
     }
 
     @Override
-    protected Path extractPath() {
-        if (finished())
-            return bestPath.extract();
-
-        return bestPath;
+    protected AStarEntry createStartEntry(int node, double weight, boolean reverse) {
+        double heapWeight = weight + weightApprox.approximate(node, reverse);
+        return new AStarEntry(EdgeIterator.NO_EDGE, node, heapWeight, weight);
     }
 
     @Override
-    protected double getCurrentFromWeight() {
-        return currFrom.weight;
+    protected AStarEntry createEntry(EdgeIteratorState edge, double weight, AStarEntry parent, boolean reverse) {
+        int neighborNode = edge.getAdjNode();
+        double heapWeight = weight + weightApprox.approximate(neighborNode, reverse);
+        AStarEntry entry = new AStarEntry(edge.getEdge(), neighborNode, heapWeight, weight);
+        entry.parent = parent;
+        return entry;
     }
 
     @Override
-    protected double getCurrentToWeight() {
-        return currTo.weight;
+    protected void updateEntry(AStarEntry entry, EdgeIteratorState edge, double weight, AStarEntry parent, boolean reverse) {
+        entry.edge = edge.getEdge();
+        entry.weight = weight + weightApprox.approximate(edge.getAdjNode(), reverse);
+        entry.weightOfVisitedPath = weight;
+        entry.parent = parent;
     }
 
     @Override
-    protected boolean finished() {
-        if (finishedFrom || finishedTo)
-            return true;
-
-        // using 'weight' is important and correct here e.g. approximation can get negative and smaller than 'weightOfVisitedPath'
-        return currFrom.weight + currTo.weight >= bestPath.getWeight();
+    protected AStarEntry getParent(AStarEntry entry) {
+        return entry.getParent();
```

Comment:
This is needed to satisfy the type system. The abstract method this method implements has return type `T extends SPTEntry` and is used to assign to a variable with the same type via 

    `entry = getParent(entry)` 

[here](https://github.com/ammagamma/graphhopper/blob/b273a31c8dd726b643696c21dcbde9684da067dc/core/src/main/java/com/graphhopper/routing/GenericDijkstraBidirection.java#L232). Calling 

   `entry = entry.getParent()`

would be a type error because we are assigning `SPTEntry` to `T` and we would need an unchecked cast like this:

    `entry = (T) entry.getParent()`

This cast is avoided by implementing `getParent` for the specific subclasses of `SPTEntry` `AStarEntry` etc.

Reply:
Hmmh, this would be an extra call just for the broken generic handling in Java. Highly likely the JIT optimizes this away but the only reason we do this is that we want to avoid this cast:

((AStarEntry) entry).weightOfVisitedPath = weight;

in GenericDijkstraBidirection in AStarBidirection. For this reason we create the separate GenericDijkstraBidirection class, add a generic argument to it and 3 getParent methods. I would say to simplify this (remove the 3 methods and the separate GenericDijkstraBidirection class) and avoid the extra getParent call we should accept this cast, which is also pretty save because the createEntry methods control object creation. If you agree I could also push my changes here.

Reply:
Sure if you think you have a better solution push it! Its also true that we need to cast at some point anyway the way it is now (the getParent() implementations all do a cast). Not sure which statment `weightOfVisitePath = weight` you mean but I guess it becomes clear to me when you push.

Reply:
Regarding JIT, yes thats what I would assume as well, but in the end you never know :) I ran the measurement class a few times for this PR and master and could not note a clear difference (but it is also hard to tell because each measurement run gives somewhat different results)